Implement UPnP support without dotNET

Yosshi999 opened the pull request:
## Introduction
Implement UPnP forwarding by using socket and XML module instead of .NET bindings.

### Relevant issues
Solve #3216

## Backwards compatibility
.NET is no longer used for UPnP

## Tests
Tested on WSL (Ubuntu 18.04.2 LTS) and a user was able to connect to the server from external.

Yosshi999 committed:
Implement native UPnP support without .NET

Yosshi999 committed:
more detailed error message in failing to use UPnP

dktapps commented on src/pocketmine/network/upnp/UPnP.php:
```diff
@@ -19,63 +19,195 @@
+		if($httpCode !== 200){
+			throw new \RuntimeException("Unable to access XML: {$response}");
+		}
+		$root = new \SimpleXMLElement($response);
```
A dependency on `ext-simplexml` needs to be added to `composer.json` for this.

dktapps replied:
This is going to spew a bunch of warnings and an `Exception` if the data isn't valid, which might crash the server. We want a nice error to be logged, we don't want a crash.

dktapps commented:
nice, this looks interesting! I can't test this myself currently, but maybe we could get tester feedback from MacOS and Windows users.

I'll review this in closer detail later.

matcracker reviewed (commented):
Question: why the function names are not camelCase?

dktapps commented:
blame @shoghicp

Yosshi999 committed:
use camelCase

Yosshi999 committed:
UPnP module depends on ext-simplexml

dktapps commented on src/pocketmine/network/upnp/UPnP.php:
```diff
@@ -138,22 +138,22 @@ public static function Setup() : void{
 	}
 
-	public static function PortForward(int $port) : void{
+	public static function portForward(int $port) : void{
```
This is BC-breaking.

dktapps commented on src/pocketmine/network/upnp/UPnP.php:
```diff
@@ -182,7 +182,7 @@ public static function PortForward(int $port) : void{
 	}
 
-	public static function RemovePortForward(int $port) : bool{
+	public static function removePortForward(int $port) : bool{
```
This is also BC-breaking.

Yosshi999 committed:
avoid BC breaking

SOF3 commented on src/pocketmine/network/upnp/UPnP.php:
```diff
@@ -19,63 +19,195 @@
+use const IP_MULTICAST_LOOP;
+use const IP_MULTICAST_TTL;
+use const SOL_SOCKET;
+use const SO_RCVTIMEO;
```
Add the corresponding function and class imports too

Yosshi999 committed:
import class/function for UPnP

dktapps commented on src/pocketmine/network/upnp/UPnP.php:
```diff
@@ -19,63 +19,201 @@
+		$location = trim($matches[1]);
+		$url = parse_url($location);
+		$response = Internet::getURL($location, 3, [], $err, $headers, $httpCode);
+		if(!$response){
```
Prefer `response === false` check here.

dktapps commented on src/pocketmine/network/upnp/UPnP.php:
```diff
@@ -19,63 +19,201 @@
+			throw new \RuntimeException("Unable to access XML: {$response}");
+		}
+		$root = new \SimpleXMLElement($response);
+		$device = $root->device;
```
This might not exist, or it might be of the wrong type.

Yosshi999 replied:
So should I check the existence of `device` by using `isset()`?

dktapps replied:
Yes, and the type.

Yosshi999 replied:
I found `SimpleXMLElement::xpath`, which enable us to explore inside a XML without executing null check or type check in each XML node (All we need will be just checking `empty($root->xpath(...))` is true or not, I think.)
I'm now rewriting the code by using this xpath so it will take a little longer to commit.

jasonwynn10 replied:
`empty()` has some issues with phpstan. It's advised to use `is_null()` where applicable for more accurate analysis.

dktapps replied:
`is_null()` is a useless function, and `isset()` has the same problems as `empty()`, so I'm a little confused what the point is here.

dktapps replied:
To be clear, the reason `empty()` is banned by PHPStan is because it disguises an `isset() and != null` check behind a misleading name.

Yosshi999 replied:
I mean, since xpath() returns false if some error occurs and returns array(0) if nothing is found, I tried to use empty() to check these two cases.
Now I'm checking them by `$xpathResult === false` and `count($xpathResult) === 0`

dktapps replied:
Yes, don't use weak type checks. That goes for `!$result` on functions that return `something|false` as well.

dktapps commented on src/pocketmine/network/upnp/UPnP.php:
```diff
@@ -19,63 +19,201 @@
+		|| !self::getChildService($device, "urn:schemas-upnp-org:service:WANIPConnection:1", $service)){
+			throw new \RuntimeException("Your router does not support portforwarding");
+		}
+		$controlURL = (string)$service->controlURL;
```
This might not exist, or it might be of the wrong type.

dktapps commented on src/pocketmine/network/upnp/UPnP.php:
```diff
@@ -19,63 +19,201 @@
 
 abstract class UPnP{
+	/** @var string */
+	private static $serviceURL;
```
This should be `string|null` with a default of `null`.

dktapps replied:
please provide a `null` default, it'll be needed for PHP 7.4.

dktapps commented on src/pocketmine/network/upnp/UPnP.php:
```diff
@@ -19,63 +19,201 @@
+	 * @return bool
+	 */
+	private static function getChildDevice(\SimpleXMLElement $device, string $deviceType, &$foundDevice) : bool{
+		foreach($device->deviceList->device as $child){
```
Again, no guarantee that `deviceList` will be of the correct type, or that it'll have a `device` member.

dktapps commented on src/pocketmine/network/upnp/UPnP.php:
```diff
@@ -19,63 +19,201 @@
+	 */
+	private static function getChildDevice(\SimpleXMLElement $device, string $deviceType, &$foundDevice) : bool{
+		foreach($device->deviceList->device as $child){
+			if((string)$child->deviceType === $deviceType){
```
No guarantee of presence or type here.

dktapps commented on src/pocketmine/network/upnp/UPnP.php:
```diff
@@ -19,63 +19,201 @@
+	 * @return bool
+	 */
+	private static function getChildService(\SimpleXMLElement $device, string $serviceType, &$foundService) : bool{
+		foreach($device->serviceList->service as $child){
```
Same as above - no guarantee of type or field presence.

dktapps commented on src/pocketmine/network/upnp/UPnP.php:
```diff
@@ -19,63 +19,201 @@
+	 */
+	private static function getChildService(\SimpleXMLElement $device, string $serviceType, &$foundService) : bool{
+		foreach($device->serviceList->service as $child){
+			if((string)$child->serviceType === $serviceType){
```
Same as above.

dktapps commented on src/pocketmine/network/upnp/UPnP.php:
```diff
@@ -19,63 +19,201 @@
+		$socket = socket_create(AF_INET, SOCK_DGRAM, SOL_UDP);
+		socket_set_option($socket, SOL_SOCKET, SO_RCVTIMEO, array("sec"=>3,"usec"=>0));
+		$contents =
+			"M-SEARCH * HTTP/1.1\r\n".
```
Formatting - there should be whitespace between the string and the concat operator.

dktapps commented on src/pocketmine/network/upnp/UPnP.php:
```diff
@@ -19,63 +19,201 @@
+		socket_set_option($socket, SOL_SOCKET, SO_RCVTIMEO, array("sec"=>3,"usec"=>0));
+		$contents =
+			"M-SEARCH * HTTP/1.1\r\n".
+			"MX: 2\r\n".
```
as above

dktapps commented on src/pocketmine/network/upnp/UPnP.php:
```diff
@@ -19,63 +19,201 @@
+		$contents =
+			"M-SEARCH * HTTP/1.1\r\n".
+			"MX: 2\r\n".
+			"HOST: 239.255.255.250:1900\r\n".
```
as above

dktapps commented on src/pocketmine/network/upnp/UPnP.php:
```diff
@@ -19,63 +19,201 @@
+			"M-SEARCH * HTTP/1.1\r\n".
+			"MX: 2\r\n".
+			"HOST: 239.255.255.250:1900\r\n".
+			"MAN: \"ssdp:discover\"\r\n".
```
as above.

dktapps commented on src/pocketmine/network/upnp/UPnP.php:
```diff
@@ -19,63 +19,201 @@
+	
+	public static function setup() : void{
+		$socket = socket_create(AF_INET, SOCK_DGRAM, SOL_UDP);
+		socket_set_option($socket, SOL_SOCKET, SO_RCVTIMEO, array("sec"=>3,"usec"=>0));
```
prefer the short array syntax `[]`, and also please fix the whitespace.

dktapps commented on src/pocketmine/network/upnp/UPnP.php:
```diff
@@ -19,63 +19,201 @@
+		socket_sendto($socket, $contents, strlen($contents), 0, "239.255.255.250", 1900);
+		socket_recvfrom($socket, $buffer, 1024, 0, $responseHost, $responsePort);
+
+		if(!$buffer || !preg_match('/location\s*:\s*(.+)\n/i', $buffer, $matches)){
```
prefer an `=== false` check here.

dktapps commented on src/pocketmine/network/upnp/UPnP.php:
```diff
@@ -19,63 +19,201 @@
+			"HOST: 239.255.255.250:1900\r\n".
+			"MAN: \"ssdp:discover\"\r\n".
+			"ST: upnp:rootdevice\r\n\r\n";
+		socket_sendto($socket, $contents, strlen($contents), 0, "239.255.255.250", 1900);
```
This might emit E_WARNING, which will turn into `ErrorException` that will crash the server.

dktapps commented on src/pocketmine/network/upnp/UPnP.php:
```diff
@@ -19,63 +19,201 @@
+			"MAN: \"ssdp:discover\"\r\n".
+			"ST: upnp:rootdevice\r\n\r\n";
+		socket_sendto($socket, $contents, strlen($contents), 0, "239.255.255.250", 1900);
+		socket_recvfrom($socket, $buffer, 1024, 0, $responseHost, $responsePort);
```
same here

dktapps commented on src/pocketmine/network/upnp/UPnP.php:
```diff
@@ -19,63 +19,201 @@
+			throw new \RuntimeException("Unable to find the router. Ensure that network discovery is enabled in Control Panel.");
+		}
+		$location = trim($matches[1]);
+		$url = parse_url($location);
```
`parse_url()` may return false if the match is invalid

dktapps commented on src/pocketmine/network/upnp/UPnP.php:
```diff
@@ -19,63 +19,201 @@
+		}
+		$root = new \SimpleXMLElement($response);
+		$device = $root->device;
+		if(!((string)$device->deviceType === "urn:schemas-upnp-org:device:InternetGatewayDevice:1")
```
This might not exist, or it might have the wrong type.

dktapps commented on src/pocketmine/network/upnp/UPnP.php:
```diff
@@ -19,63 +19,201 @@
+		return false;
+	}
+	
+	public static function setup() : void{
```
I think it would make more sense if this was named `getControlUrl()` and returned `string`. Then have it use the static field as a cache.

dktapps commented on src/pocketmine/network/upnp/UPnP.php:
```diff
@@ -19,63 +19,201 @@
+				'<NewEnabled>1</NewEnabled>'.
+				'<NewPortMappingDescription>PocketMine-MP</NewPortMappingDescription>'.
+				'<NewLeaseDuration>0</NewLeaseDuration>'.
+			'</u:AddPortMapping>';
```
Formatting here needs whitespace between strings and concat operators.

dktapps commented on src/pocketmine/network/upnp/UPnP.php:
```diff
@@ -19,63 +19,201 @@
-		$myLocalIP = Internet::getInternalIP();
+		$contents =
+			'<?xml version="1.0"?>'.
+			'<s:Envelope xmlns:s="http://schemas.xmlsoap.org/soap/envelope/" s:encodingStyle="http://schemas.xmlsoap.org/soap/encoding/">'.
```
Needs space between string and concat operators (this line and the one above).

dktapps commented on src/pocketmine/network/upnp/UPnP.php:
```diff
@@ -19,63 +19,201 @@
+				'<NewRemoteHost></NewRemoteHost>'.
+				'<NewExternalPort>' . $port . '</NewExternalPort>'.
+				'<NewProtocol>UDP</NewProtocol>'.
+			'</u:DeletePortMapping>';
```
whitespace

dktapps commented on src/pocketmine/network/upnp/UPnP.php:
```diff
@@ -19,63 +19,201 @@
+
+		$contents =
+			'<?xml version="1.0"?>'.
+			'<s:Envelope xmlns:s="http://schemas.xmlsoap.org/soap/envelope/" s:encodingStyle="http://schemas.xmlsoap.org/soap/encoding/">'.
```
whitespace

dktapps reviewed (changes requested):
I've left a few comments, mostly about validation and formatting.

Yosshi999 committed:
[WIP] Use xpath instead(Untested), better formatting

dktapps commented on src/pocketmine/network/upnp/UPnP.php:
```diff
@@ -19,63 +19,183 @@
+			socket_sendto($socket, $contents, strlen($contents), 0, "239.255.255.250", 1900);
+			socket_recvfrom($socket, $buffer, 1024, 0, $responseHost, $responsePort);
+			socket_close($socket);
+		}catch(\Exception $e){
```
Not a fan of global `Exception` caught here.

Yosshi999 replied:
I need to put `set_error_handler([Utils::class, 'errorExceptionHandler']);` before try and catch `ErrorException` and put `restore_error_handler();` in finally, right?

dktapps replied:
No, it's already set by default. That said, it's better to use @ and check for false returns, because static analysers can't make sense of error handlers.

dktapps commented on src/pocketmine/network/upnp/UPnP.php:
```diff
@@ -19,63 +19,183 @@
+			// if result is false or [], i.e. error or notfound
+			throw new \RuntimeException("Your router does not support portforwarding");
+		}
+		$controlURL = (string)($xpathResult[0]);
```
is this guaranteed to be a string at the type level?

Yosshi999 replied:
xpath() returns an array of SimpleXMLElement, and SimpleXMLElement has __toString() so I think casting to string always works.

dktapps commented on src/pocketmine/network/upnp/UPnP.php:
```diff
@@ -19,63 +19,183 @@
-		if(!class_exists("COM")){
-			throw new \RuntimeException("UPnP requires the com_dotnet extension");
+
+		if(!isset(self::$serviceURL)){
```
please use a `!== null` check here, `isset()` should be avoided where possible.

dktapps commented on src/pocketmine/network/upnp/UPnP.php:
```diff
@@ -19,63 +19,183 @@
 			return false;
 		}
-		if(Utils::getOS() != "win" or !class_exists("COM")){
+		if(!isset(self::$serviceURL)){
```
null-check here please

dktapps commented on src/pocketmine/network/upnp/UPnP.php:
```diff
@@ -19,63 +19,183 @@
+			throw new \RuntimeException("Your router does not support portforwarding");
+		}
+		$controlURL = (string)($xpathResult[0]);
+		$serviceURL = "{$url['host']}:{$url['port']}/{$controlURL}";
```
I'd prefer if this used concat operators or even `sprintf()`, this is difficult to read

dktapps commented on src/pocketmine/network/upnp/UPnP.php:
```diff
@@ -19,63 +19,183 @@
+		$defaultInternalError = libxml_use_internal_errors(true);
+		try{
+			$root = new \SimpleXMLElement($response);
+		}catch(\Exception $e){
```
please check a more specific exception if possible

Yosshi999 replied:
`get_class($e)` returned 'Exception', so I think it raises Exception itself.

dktapps replied:
That's fair enough. I just get uncomfortable seeing generalized catches like this.

Yosshi999 committed:
better format, consistent function name

dktapps commented on src/pocketmine/network/upnp/UPnP.php:
```diff
@@ -131,7 +137,7 @@ public static function getControlUrl() : string{
 			throw new \RuntimeException("Your router does not support portforwarding");
 		}
 		$controlURL = (string)($xpathResult[0]);
```
```suggestion
		$controlURL = (string) $xpathResult[0];
```

Yosshi999 committed:
Remove unnecessary parentheses

Co-authored-by: Dylan T. <odigiman@gmail.com>

Yosshi999 commented:
Unfortunately my network environment has been changed and I can't access the UPnP setting of my router, so I can't test the current implementation. I'd like someone to help for testing this.
After making sure that it works, I will squash the commits since they are messy.

dktapps commented:
don't worry too much about squashing, I can squash merge :slight_smile:

Yosshi999 committed:
rebase stable and update content-hash

Yosshi999 committed:
poor rebasing; forgot to delete dot-com test

Yosshi999 committed:
resolve phpstan errors

dktapps commented on src/pocketmine/network/upnp/UPnP.php:
```diff
@@ -19,63 +19,232 @@
+		socket_close($socket);
+		$pregResult = preg_match('/location\s*:\s*(.+)\n/i', $buffer, $matches);
+		if($pregResult === false){
+			// you can use preg_last_error_msg() since PHP 8.
```
```suggestion
			//TODO: replace with preg_last_error_msg() in PHP 8.
```

dktapps commented on src/pocketmine/network/upnp/UPnP.php:
```diff
@@ -19,63 +19,232 @@
+				default:
+					$errormsg = "unknown error";
+			}
+			throw new \RuntimeException("Preg error: " . $errormsg);
```
This would take much less space if you do something like this: https://github.com/pmmp/PocketMine-MP/blob/15401d740f11f3a543abaecc7a449bcba3c2fc91/src/pocketmine/utils/TextFormat.php#L77-L88

dktapps commented on src/pocketmine/network/upnp/UPnP.php:
```diff
@@ -19,63 +19,232 @@
+			'/controlURL'
+		);
+		if($xpathResult === false || count($xpathResult) === 0){
+			// if result is false or [], i.e. error or notfound
```
```suggestion
```

dktapps commented on composer.json:
```diff
@@ -23,6 +23,7 @@
       "ext-yaml": ">=2.0.0",
       "ext-zip": "*",
       "ext-zlib": ">=1.2.11",
+      "ext-simplexml": "*",
```
The extensions should be alphabetically ordered.

dktapps reviewed (commented):
I can't test this right now due to network configuration.

Yosshi999 committed:
Apply suggestions for better comments

Co-authored-by: Dylan T. <odigiman@gmail.com>

Yosshi999 committed:
apply suggestions

dktapps committed:
Merge branch 'stable' into UPnP-without-dotNET

dktapps commented:
This appears to be completely dysfunctional. The `xpath()` call is not able to find even basic elements like `//device`, despite them obviously being present.

dktapps committed:
Fixed xpath query

dktapps committed:
shut up stan

dktapps committed:
...

dktapps committed:
Merge branch 'stable' into UPnP-without-dotNET

dktapps committed:
Imports cleanup

dktapps commented:
This is now working correctly on my Hub 3.0.

dktapps committed:
Make router discovery broadcast more fault-tolerant
this might fail for a number of reasons:
1) the packet got lost (it's UDP)
2) the socket picked up some noise from somewhere else on the network
Both of these errors are non-fatal.

dktapps committed:
UPnP is no longer limited to Windows
